tests/federation_room_join_partial_state_test.go

```diff
@@ -3295,6 +3295,90 @@ func TestPartialStateJoin(t *testing.T) {
 		}
 	})
 
+	t.Run("User directory is correctly updated once state re-sync completes", func(t *testing.T) {
+		rocky := deployment.RegisterUser(t, "hs1", "rocky", "pass", false)
+
+		server := createTestServer(t, deployment)
+		cancel := server.Listen()
+		defer cancel()
+		serverRoom := createTestRoom(t, server, rocky.GetDefaultRoomVersion(t))
+
+		// add some new users to avoid the test being polluted by previous tests
+		serverRoom.AddEvent(createJoinEvent(t, server, serverRoom, server.UserID("rod")))
+		serverRoom.AddEvent(createJoinEvent(t, server, serverRoom, server.UserID("todd")))
+
+		// start a partial state join
+		psjResult := beginPartialStateJoin(t, server, serverRoom, rocky)
+		defer psjResult.Destroy(t)
+
+		// sanity check - before the state has completed syncing state we would expect rocky to show up
+		reqBody := client.WithJSONBody(t, map[string]interface{}{
+			"search_term": rocky.UserID,
+		})
+		time.Sleep(time.Second * 1)
+		res := rocky.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "user_directory", "search"}, reqBody)
+		must.MatchResponse(t, res, match.HTTPResponse{
+			StatusCode: 200,
+			JSON: []match.JSON{
+				match.JSONKeyEqual("results.0.user_id", "@rocky:hs1"),
+			}})
+
+		// .. but not rod's
+		reqBody2 := client.WithJSONBody(t, map[string]interface{}{
+			"search_term": "rod",
+		})
+		res2 := rocky.MustDoFunc(t, "POST", []string{"_matrix", "client", "v3", "user_directory", "search"}, reqBody2)
+		must.MatchResponse(t, res2, match.HTTPResponse{
+			StatusCode: 200,
+			JSON: []match.JSON{
+				match.JSONKeyEqual("results", [0]string{}),
+			}})
+
+		// finish syncing the state
+		psjResult.FinishStateRequest()
+		awaitPartialStateJoinCompletion(t, psjResult.ServerRoom, rocky)
+
+		// the rooms stats are updated by a background job in Synapse which is not guaranteed to have completed by the time
+		// the state sync has completed. We check for up to 3 seconds that the job has completed, after which Charlie
+		// the job should have finished and Charlie and Derek should be visible in the user directory
+		time.Sleep(time.Second * 3)
+
+		reqBody3 := client.WithJSONBody(t, map[string]interface{}{
+			"search_term": "rod",
+		})
+		rodFullId := "@rod:" + server.ServerName()
```

Comment:
We would usually write this as `server.UserID("rod")`
and the same for todd below.